### ISSUE TITLE
error: 'SO_REUSEPORT' was not declared in this scope

### ISSUE THREAD

**Issue by nmilosev** [pytorch/pytorch#29093]
## 🐛 Bug

When building from master, the error in the title appears.

## To Reproduce

1. checkout master
1. build

## Expected behavior

Build works

## Environment

 - PyTorch Version: 1.4 master
 - OS (e.g., Linux): Fedora Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): `python setup.py develop`
 - Python version: 3.7.4
 - CUDA/cuDNN version: irrelevant
 - GPU models and configuration: irrelevant
 - Any other relevant information: irrelevant

## Additional context

Problem is with nccl, as described here: https://github.com/NVIDIA/nccl/issues/244

The patch described there seems to be working but a workaround would be nice.

**Comment by soumith** [comment]
cc: @mrshenli @pietern a nccl upgrade might fix it because NVIDIA/nccl#244

**Comment by pietern** [comment]
@nmilosev Can you confirm your kernel version?

The issue suggests this is a problem only on old Linux kernel, so I want to make sure this is the case for you as well before doing a NCCL upgrade. A NCCL upgrade also means we're no longer on a released version (we're on 2.4.8 today), since this fix is not part of an official release.

**Comment by nmilosev** [comment]
I tried both on Fedora 31:

```
Linux nidavellir 5.3.7-301.fc31.x86_64 #1 SMP Mon Oct 21 19:18:58 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

and on CentOS 7:

```
Linux axiom 3.10.0-862.6.3.el7.x86_64 #1 SMP Tue Jun 26 12:13:22 CDT 2018 x86_64 x86_64 x86_64 GNU/Linux
```

Same output :(

**Comment by pietern** [comment]
The man page for `socket(7)` says that `SO_REUSEPORT` was added in Linux 3.9, so you should be fine with both of those kernels. Can you `grep` for the constant in `/usr/include` and see what happens?

Perhaps both of them are using older kernel headers?

**Comment by nmilosev** [comment]
Of course:

```
$ grep -Ril 'SO_REUSEPORT' /usr/include/                                                                                                                      
/usr/include/asm-generic/socket.h
/usr/include/boost/asio/detail/impl/socket_ops.ipp
```

But, I am building inside `conda` with gcc7 from conda, so you probably want this:

```
grep -Ril 'SO_REUSEPORT' ~/miniconda3/envs/pytorchcontrib/include/
```

So is this the issue? :)

I by default have gcc 4.8.5 which is too old. On Fedora I have gcc 9.x which is too new. So I want to use gcc 7 which I got with `conda install -c conda-forge gxx_linux-64`.

**Comment by pietern** [comment]
Yeah, I think this is it, because `conda` uses its own system root.

You can look for the path at `x86_64-conda_cos6-linux-gnu`, which includes its own `asm-generic/socket.h`. These files are provided by 4.8.5 and friends, so it must be an old version.

We run conda builds for PyTorch but don't use the conda compilers.

**Comment by nmilosev** [comment]
@pietern That's the issue, look at this:

```
$ grep -Ril 'SO_REUSEPORT' ~/miniconda3/envs/pytorchcontrib/                                                                         
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/asyncio/base_events.py                                                                         
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/asyncio/events.py                                                                              
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/asyncio/__pycache__/events.cpython-37.pyc                                                      
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/asyncio/__pycache__/base_events.cpython-37.pyc                                                 
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/test/support/__init__.py                                                                       
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/test/support/__pycache__/__init__.cpython-37.pyc                                               
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/site-packages/mypy/typeshed/stdlib/2/_socket.pyi                                               
/home/milosevicn/miniconda3/envs/pytorchcontrib/lib/python3.7/site-packages/mypy/typeshed/stdlib/2and3/socket.pyi
```

Nowhere to be found. Sadly I cannot use my system gcc because of the mentioned issues. I will look into what are my options in resolving this (maybe another conda package with that header).

Thanks for your help, I think this is safe to close?

**Comment by nmilosev** [comment]
Building with GCC7 from devtoolkit-7 (scl) works (with CUDA, no special environment). 

I think we should recommend at least to our users what to compile pytorch with.

**Comment by pietern** [comment]
Detecting this situation, with a newer GCC and an older sysroot, is tricky though.

The README doesn't say anything specific about the compiler to use when compiling from source, except that it must be capable of compiling C++14 source code. We could include a note explicitly saying that the default conda compiler and sysroot are too old.

**Comment by SsnL** [comment]
I'm also seeing this problem with conda gcc 7.3, which is what our README suggest using...

**Comment by pietern** [comment]
@SsnL Can you confirm the sysroot for conda gcc 7.3?

**Comment by jlquinn** [comment]
I'm running into this issue trying to compile v1.4 branch.  CentOS 7, conda 4.8.3, conda gcc 7.3.0, conda cuda toolkit 10.1.

I've tried USE_SYSTEM_NCCL, USE_STATIC_NCCL, installing conda nccl.  So far nothing works.

**Comment by YingleiZhang** [comment]
I think NCCL had it fix a few month ago, and if you upgrade the NCCL version for PyTorch source code, it would fix the issue. 
I cut a ticket for similar issue: https://github.com/pytorch/pytorch/issues/35363

**Comment by bryan-lunt** [comment]
What do we do if we're stuck on ancient supercomputers and have to use particular versions of cudatoolkit, implying specific versions of NCCL?

**Comment by rgommers** [comment]
> What do we do if we're stuck on ancient supercomputers and have to use particular versions of cudatoolkit, implying specific versions of NCCL?

@bryan-lunt NCCL is shipped with PyTorch in `third_party/nccl`. Applying the diff of https://github.com/NVIDIA/nccl/commit/7f2b337e703d73ed369937c9996e1f3d5f664ad0 to that will work with `cudatoolkit` in a conda env.

**Comment by rgommers** [comment]
> The README doesn't say anything specific about the compiler to use when compiling from source, except that it must be capable of compiling C++14 source code. We could include a note explicitly saying that the default conda compiler and sysroot are too old.

I'm not sure that's helpful. Compiler is GCC 7.3, while PyTorch CI still has GCC 5, so that's not an issue. And the conda sysroot is hard to avoid. Also, using conda compilers + cudatoolkit is the best way to get a completely consistent environment often (from "ancient supercomputer" as above to, e.g. Arch Linux which often has the very latest GCC that doesn't work with PyTorch master yet).

It's not like there's a lot of fixes needed for conda compilers + sysroot, here's the current status: https://github.com/Quansight/pearu-sandbox/blob/master/working-envs/activate-pytorch-dev.sh

Adding an all-conda CI job would be much better, would save a lot of friction.
